File: Data Structures & Algorithms/valid-binary-search-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool isValidBSTDFS(TreeNode* root, int left, int right) {
        if (!root) return true;
        if (root->val <= left || root->val >= right) return false;
        return (
            isValidBSTDFS(root->left, left, root->val) && 
            isValidBSTDFS(root->right, root->val, right)
        );
    }

    bool isValidBST(TreeNode* root) {
        return isValidBSTDFS(root, INT_MIN, INT_MAX);
    }
};

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxDepthDFS(self, root: Optional[TreeNode], is_balanced: List[bool]) -> int:
        if not root:
            return 0
        left = self.maxDepthDFS(root.left, is_balanced)
        right = self.maxDepthDFS(root.right, is_balanced)
        is_balanced[0] &= (abs(left - right) <= 1)
        return max(left, right) + 1

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        is_balanced = [True]
        self.maxDepthDFS(root, is_balanced)
        return is_balanced[0]
        